### ISSUE TITLE
Scale problems with 10x count.wdl

### ISSUE THREAD

**Issue by ambrosejcarr** [HumanCellAtlas/skylab#26]
I'm running into the following problem (weird formatting, the `success` arrays are in the same level in the parent `vector`: 

```
Vector(Success(
    ["gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-extract_reads_split/glob-c9856fc57cfd1c1bbcfeac008610354e/partition_aa",
     "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-extract_reads_split/glob-c9856fc57cfd1c1bbcfeac008610354e/partition_ab",
     "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-extract_reads_split/glob-c9856fc57cfd1c1bbcfeac008610354e/partition_ac",
     "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-extract_reads_split/glob-c9856fc57cfd1c1bbcfeac008610354e/partition_ad",
     "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-extract_reads_split/glob-c9856fc57cfd1c1bbcfeac008610354e/partition_ae",
     "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-extract_reads_split/glob-c9856fc57cfd1c1bbcfeac008610354e/partition_af"]),
       Success([
           ["gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-0/glob-1b7acead3c118bfa3530813e3d9b058b/I1-0000.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-0/glob-1b7acead3c118bfa3530813e3d9b058b/I1-0001.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-0/glob-1b7acead3c118bfa3530813e3d9b058b/I1-0002.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-0/glob-1b7acead3c118bfa3530813e3d9b058b/R1-0000.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-0/glob-1b7acead3c118bfa3530813e3d9b058b/R1-0001.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-0/glob-1b7acead3c118bfa3530813e3d9b058b/R1-0002.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-0/glob-1b7acead3c118bfa3530813e3d9b058b/R2-0000.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-0/glob-1b7acead3c118bfa3530813e3d9b058b/R2-0001.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-0/glob-1b7acead3c118bfa3530813e3d9b058b/R2-0002.fastq"],
           ["gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-1/glob-1b7acead3c118bfa3530813e3d9b058b/I1-0000.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-1/glob-1b7acead3c118bfa3530813e3d9b058b/I1-0001.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-1/glob-1b7acead3c118bfa3530813e3d9b058b/I1-0002.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-1/glob-1b7acead3c118bfa3530813e3d9b058b/R1-0000.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-1/glob-1b7acead3c118bfa3530813e3d9b058b/R1-0001.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-1/glob-1b7acead3c118bfa3530813e3d9b058b/R1-0002.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-1/glob-1b7acead3c118bfa3530813e3d9b058b/R2-0000.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-1/glob-1b7acead3c118bfa3530813e3d9b058b/R2-0001.fastq",
            "gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/call-chunk_reads_main/shard-1/glob-1b7acead3c118bfa3530813e3d9b058b/R2-0002.fastq"
           ]]))
Requires exactly two evaluated array values of equal length.
```

The exception occurs immediately following the successful completion of:
```
2017-09-30 16:55:23,507 cromwell-system-akka.dispatchers.backend-dispatcher-269 INFO  - JesAsyncBackendJobExecutionActor [UUID(a6dde172)count.extract_reads_split:NA:1]: Status change from Running to Success
2017-09-30 16:55:24,797 cromwell-system-akka.dispatchers.engine-dispatcher-343 ERROR - WorkflowManagerActor Workflow a6dde172-0bf7-41b1-8524-c6f513427868 failed (during ExecutingWorkflowState):
```

Suggesting the problem is in the following `count.wdl` code:

```
  scatter(split_pair in zip(extract_reads_split.chunks, chunk_reads_main.fastq_chunks)) {
      call extract_reads_main {
        input:
          out_chunks = chunk_reads_join.out_chunks,
          barcode_whitelist = setup_chunks.barcode_whitelist,
          primers = primers,
          reads_per_file = reads_per_file,
          subsample_rate = subsample_rate,
          chemistry_def = setup_chunks.chemistry_def,
          chunk = split_pair.left,
          fastq_chunks = split_pair.right
    }
  }
```

It seems like the `chunk-reads-main` array is being sharded, but the `extract-reads-split` operation is not, leading to the `zip` problem. I've reproduced this error on datasets with 10 million and 40 million reads. 

@dshiga @mckinsel Do either of you have any insights into this discrepancy, or have suggestions about who I could ask about this?

**Comment by mckinsel** [comment]
Hey can you point me to the whole cromwell-execution directory? My initial guess is that we're running into `reads_per_file` assumptions being violated, but I'm not sure. I doubt it's a cromwell/JES thing right now.

**Comment by ambrosejcarr** [comment]
execution directory: `gs://broad-dsde-mint-dev-cromwell-execution/count/a6dde172-0bf7-41b1-8524-c6f513427868/`

**Comment by ambrosejcarr** [comment]
This was identified as a bug with `reads_per_file`. For now, we can work around the problem by setting this value to a very large number.

**Comment by mckinsel** [comment]
A proposed fix is in #27

**Comment by ambrosejcarr** [comment]
Fix got me further; I forgot to change my `reads_per_file` number back to something smaller. It turns out `10,000,000,000` is not a good idea. Here's the traceback for `call-chunk_reads_main`

```
Could not decode '-727379968' to usize for '--reads-per-fastq'.
Traceback (most recent call last):
File "/martian/adapters/python/martian_shell.py", line 524, in _main
stage.main()
File "/martian/adapters/python/martian_shell.py", line 490, in main
self._run(lambda: self._module.main(args, outs))
File "/martian/adapters/python/martian_shell.py", line 459, in _run
cmd()
File "/martian/adapters/python/martian_shell.py", line 490, in 
self._run(lambda: self._module.main(args, outs))
File "/cellranger/mro/stages/common/chunk_reads/__init__.py", line 53, in main
subprocess.check_call(chunk_reads_args)
File "/usr/lib/python2.7/subprocess.py", line 186, in check_call
raise CalledProcessError(retcode, cmd)
CalledProcessError: Command '['chunk_reads', '--reads-per-fastq', '-727379968', 'files/', 'fastq_chunk', '--martian-args', 'chunk_args.json']' returned non-zero exit status 1
jq: error (at out_chunks.json:1): Cannot iterate over null (null)
ln: failed to access 'fastq_chunk_*.tar.gz': No such file or directory
```

Unprotected Integer overflows! What *fun*. 

I'm re-running with `reads_per_file: 5,000,000`.

**Comment by ambrosejcarr** [comment]
@mckinsel I managed to push the complete `pbmc8k` dataset up to `extract_reads_split`, but am running into a `jq` error that I'm not best positioned to fix (you're the expert!): 

`extract_reads_split` is failing with the error (lines after first are likely side-effects of the error):

```
/cromwell_root/exec.sh: line 11: /usr/bin/jq: Argument list too long
No JSON object could be decodedTraceback (most recent call last):
File "/martian/adapters/python/martian_shell.py", line 524, in _main
stage.main()
File "/martian/adapters/python/martian_shell.py", line 482, in main
self._run(lambda: self._record_result(
File "/martian/adapters/python/martian_shell.py", line 459, in _run
cmd()
File "/martian/adapters/python/martian_shell.py", line 483, in 
lambda: self._module.split(args)))
File "/martian/adapters/python/martian_shell.py", line 463, in _record_result
self._result = cmd()
File "/martian/adapters/python/martian_shell.py", line 483, in 
lambda: self._module.split(args)))
File "/cellranger/mro/stages/common/extract_reads/__init__.py", line 51, in split
gem_groups = [chunk['gem_group'] for chunk in args.chunks]
AttributeError: 'Record' object has no attribute 'chunks'
jq: error: Could not open file _stage_defs: No such file or directory
ln: failed to access 'partition_*': No such file or directory
```

line 11 is the jq command that references out_chunks.json
```
echo "$(jq --arg var "$(</cromwell_root/broad-dsde-mint-dev-cromwell-execution/count/090b7219-cb9b-40dd-ab0f-5e177053169e/call-chunk_reads_join/out_chunks.json)" \'. + {"chunks": $var | fromjson}\' _args)" > _args
```
Specifically, out_chunks.json is 6954 lines long, and maybe jq doesn't like that. 

For reference, I'm running on the `adjust_10x_resources` branch which has some changes to `count.wdl` to fix upstream scale problems; feel free to commit a bugfix there, I won't do more work on it until this is fixed. 

cromwell root: `gs://broad-dsde-mint-dev-cromwell-execution/count/090b7219-cb9b-40dd-ab0f-5e177053169e/`

**Comment by ambrosejcarr** [comment]
@mckinsel do you have the Cromwell root for the most recent run failure you ran with the updated disk and memory requirements?

**Comment by mckinsel** [comment]
Yeah it's job 9f8f6f88-a422-4ee4-aaee-28a309425c3f. It looks like something weird happened with the fastq chunk flattening.

**Comment by mckinsel** [comment]
It takes some work to find the actual failure, but this is it:

```
        "failures": [
          {
            "causedBy": [
              {
                "causedBy": [],
                "message": "Pipeline 5043316292578993923: Unable to evaluate parameters: parameter \"extract_reads_main.fastq_chunk-0\" has invalid value: 1/fastq_chunk_0038.tar.gz"
              }
            ],
            "message": "Unable to complete JES Api Request"
          }
        ],
```

**Comment by ambrosejcarr** [comment]
Cromwell root for most recent run. In-progress.
gs://broad-dsde-mint-dev-cromwell-execution/count/c9f61de6-87fa-4f4a-bbb6-7a192d80162b/

**Comment by ambrosejcarr** [comment]
New run: `gs://broad-dsde-mint-dev-cromwell-execution/count/9c945fc1-fb36-4aa9-9e8e-456d7ddbf8c2/`

**Comment by ambrosejcarr** [comment]
Something to be aware of, these big splits cause some massive download overhead. This run has been just localizing data from the previous split task for almost 3 hours. 

<img width="1679" alt="screen shot 2017-10-13 at 5 25 22 pm" src="https://user-images.githubusercontent.com/5239152/31567062-a08673e0-b03b-11e7-905d-25011df74a07.png">

**Comment by ambrosejcarr** [comment]
<img width="597" alt="screen shot 2017-10-13 at 5 48 20 pm" src="https://user-images.githubusercontent.com/5239152/31567734-e9d1d9a6-b03e-11e7-90e7-8f619ba09f1b.png">


We're past the last point of failure! Also here's what it looks like when it's complete. green = localizing, yellow = delocalizing.

**Comment by mckinsel** [comment]
This should be fixed by #30, so I'm going to mark it closed. @ambrosejcarr, feel free to open another issue about needless file localization if you think it's necessary to address.
